src/syscall_handler.c

```diff
@@ -112,6 +112,79 @@ int sys_close( int fd )
 	return ENOSYS;
 }
 
+int sys_w_color( int wd, int r, int g, int b ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    struct graphics_color c;
+    c.r = r;
+    c.g = g;
+    c.b = b;
+    c.a = 0;
+    graphics_fgcolor( current->windows[wd], c );
+    return 0;
+}
+
+int sys_w_rect( int wd, int x, int y, int w, int h ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    graphics_rect( current->windows[wd], x, y, w, h );
+    return 0;
+}
+
+int sys_w_clear( int wd, int x, int y, int w, int h ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    graphics_clear( current->windows[wd], x, y, w, h );
+    return 0;
+}
+
+int sys_w_line( int wd, int x, int y, int w, int h ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    graphics_line( current->windows[wd], x, y, w, h );
+    return 0;
+}
+
+int sys_w_char( int wd, int x, int y, char c ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    graphics_char( current->windows[wd], x, y, c );
+    return 0;
+}
+
+int sys_w_string( int wd, int x, int y, char *s ) {
+    if (wd < 0 || wd >= current->window_count) {
+        return ENOENT;
+    }
+    int i;
+    for (i = 0; s[i]; i++) {
+        graphics_char( current->windows[wd], x+i*8, y, s[i] );
+    }
+    return 0;
+}
+
+int sys_w_create( int x, int y, int w, int h ) {
+    if (current->window_count >= PROCESS_MAX_WINDOWS) {
+        return ENOENT;
+    }
+    current->windows[current->window_count] = graphics_create(&graphics_root);
+    if (!current->windows[current->window_count]) {
+        return ENOENT;
+    }
+
```

Comment:
Here is the interesting bit: When a user process requests a new window, the first argument should be a window descriptor, and the coordinates should be interpreted as `within` the parent window.  Reject the request if it doesn't fit.

Reply:
But how should a process request its first window - i.e. what is the window descriptor of the root window? Would we want to do something like have children inherit their descriptors from their parents, and have the first process have the root window open? Or would we want something like a "open root window" system call?

Reply:
Yes!  The first process begins with a single descriptor for the entire screen, and children inherit descriptors from the parent.